### DISCUSSION TITLE
Fabric8 IO client creation is failing with below stacktrace

### DISCUSSION THREAD

**Discussion by srinivasev** [fabric8io/kubernetes-client#4653] [Q&A]
Hi Team,

I am facing below error during spring boot startup. Any idea what could be the issue ?.
Below dependency is used in my spring boot project.

```
    <dependency>
      <groupId>org.springframework.cloud</groupId>
      <artifactId>spring-cloud-starter-kubernetes-fabric8-config</artifactId>
    </dependency>
```

Stack trace -

`14:36:37.314 ERROR SpringApplication 835 Application run failed io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred. at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:103) at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:97) at io.fabric8.kubernetes.client.utils.HttpClientUtils.createHttpClient(HttpClientUtils.java:253) at io.fabric8.kubernetes.client.utils.HttpClientUtils.createHttpClient(HttpClientUtils.java:85) at io.fabric8.kubernetes.client.BaseClient.<init>(BaseClient.java:53) at io.fabric8.kubernetes.client.BaseClient.<init>(BaseClient.java:45) at io.fabric8.kubernetes.client.BaseKubernetesClient.<init>(BaseKubernetesClient.java:151) at io.fabric8.kubernetes.client.DefaultKubernetesClient.<init>(DefaultKubernetesClient.java:32) at org.springframework.cloud.kubernetes.fabric8.profile.Fabric8ProfileEnvironmentPostProcessor.isInsideKubernetes(Fabric8ProfileEnvironmentPostProcessor.java:29) at org.springframework.cloud.kubernetes.commons.profile.AbstractKubernetesProfileEnvironmentPostProcessor.addKubernetesProfileIfMissing(AbstractKubernetesProfileEnvironmentPostProcessor.java:81) at org.springframework.cloud.kubernetes.commons.profile.AbstractKubernetesProfileEnvironmentPostProcessor.postProcessEnvironment(AbstractKubernetesProfileEnvironmentPostProcessor.java:66) at org.springframework.boot.env.EnvironmentPostProcessorApplicationListener.onApplicationEnvironmentPreparedEvent(EnvironmentPostProcessorApplicationListener.java:102) at org.springframework.boot.env.EnvironmentPostProcessorApplicationListener.onApplicationEvent(EnvironmentPostProcessorApplicationListener.java:87) at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:176) at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:169) at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:143) at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:131) at org.springframework.boot.context.event.EventPublishingRunListener.environmentPrepared(EventPublishingRunListener.java:85) at org.springframework.boot.SpringApplicationRunListeners.lambda$environmentPrepared$2(SpringApplicationRunListeners.java:66) at java.base/java.util.ArrayList.forEach(ArrayList.java:1541) at org.springframework.boot.SpringApplicationRunListeners.doWithListeners(SpringApplicationRunListeners.java:120) at org.springframework.boot.SpringApplicationRunListeners.doWithListeners(SpringApplicationRunListeners.java:114) at org.springframework.boot.SpringApplicationRunListeners.environmentPrepared(SpringApplicationRunListeners.java:65) at org.springframework.boot.SpringApplication.prepareEnvironment(SpringApplication.java:343) at org.springframework.boot.SpringApplication.run(SpringApplication.java:301) at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:164) at org.springframework.cloud.bootstrap.BootstrapApplicationListener.bootstrapServiceContext(BootstrapApplicationListener.java:195) at org.springframework.cloud.bootstrap.BootstrapApplicationListener.onApplicationEvent(BootstrapApplicationListener.java:114) at org.springframework.cloud.bootstrap.BootstrapApplicationListener.onApplicationEvent(BootstrapApplicationListener.java:77) at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:176) at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:169) at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:143) at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:131) at org.springframework.boot.context.event.EventPublishingRunListener.environmentPrepared(EventPublishingRunListener.java:85) at org.springframework.boot.SpringApplicationRunListeners.lambda$environmentPrepared$2(SpringApplicationRunListeners.java:66) at java.base/java.util.ArrayList.forEach(ArrayList.java:1541) at org.springframework.boot.SpringApplicationRunListeners.doWithListeners(SpringApplicationRunListeners.java:120) at org.springframework.boot.SpringApplicationRunListeners.doWithListeners(SpringApplicationRunListeners.java:114) at org.springframework.boot.SpringApplicationRunListeners.environmentPrepared(SpringApplicationRunListeners.java:65) at org.springframework.boot.SpringApplication.prepareEnvironment(SpringApplication.java:343) at org.springframework.boot.SpringApplication.run(SpringApplication.java:301) at org.springframework.boot.SpringApplication.run(SpringApplication.java:1317) at org.springframework.boot.SpringApplication.run(SpringApplication.java:1306) at com.ericsson.oss.common.service.ns.NotificationServiceApplication.main(NotificationServiceApplication.java:35) at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.base/java.lang.reflect.Method.invoke(Method.java:566) at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49) at org.springframework.boot.loader.Launcher.launch(Launcher.java:107) at org.springframework.boot.loader.Launcher.launch(Launcher.java:58) at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:88) Caused by: java.security.cert.CertificateException: Could not parse certificate: java.io.IOException: Empty input at java.base/sun.security.provider.X509Factory.engineGenerateCertificate(X509Factory.java:115) at java.base/java.security.cert.CertificateFactory.generateCertificate(CertificateFactory.java:355) at io.fabric8.kubernetes.client.internal.CertUtils.createTrustStore(CertUtils.java:98) at io.fabric8.kubernetes.client.internal.CertUtils.createTrustStore(CertUtils.java:74) at io.fabric8.kubernetes.client.internal.SSLUtils.trustManagers(SSLUtils.java:115) at io.fabric8.kubernetes.client.internal.SSLUtils.trustManagers(SSLUtils.java:91) at io.fabric8.kubernetes.client.utils.HttpClientUtils.createHttpClient(HttpClientUtils.java:147) ... 49 common frames omitted Caused by: java.io.IOException: Empty input at java.base/sun.security.provider.X509Factory.engineGenerateCertificate(X509Factory.java:111) ... 55 common frames omitted`

**Comment by shawkins** [comment]
This should be happening based upon the client attempting to read the default truststore.  This is coming from config.getCaCertData() or config.getCaCertFile(). The values for those properties is typically auto configured.  If you are not modifying them, please use new ConfigBuilder().build() to inspect the state of a fresh configuration and see what those properties contain.

**Comment by srinivasev** [comment]
Hi @shawkins ,

Thanks for the response.

By the way, one more doubt.. I checked my kubernetes cluster and the Fabric8 IO compatibility matrix - [https://github.com/fabric8io/kubernetes-client#compatibility-matrix]

Below are the findings -

K8s version = 1.24
Fabric8 IO =

<groupId>io.fabric8</groupId>
<artifactId>kubernetes-client</artifactId>
<version>5.10.2</version>

Is this incompatibility is causing above failure while creating the client ?

**Reply by shawkins** [reply]
> Is this incompatibility is causing above failure while creating the client ?

No, that affects the generated models.

**Reply by srinivasev** [reply]
Okay,
I will try to check the contents exist in the java truststore (microservice pod).
Do you think the truststore is not generated properly causing the fabric8 io failure based on above stacktrace ?

**Reply by srinivasev** [reply]
Hi @shawkins ,
Can you please share me the document which explains the flow of farbic8 io client creation ?.

I was not aware there is a call which reads the truststore before creating the fabric8 io client. Need to understand more on the same.

**Reply by manusa** [reply]
There is no document explaining the flow.
The client will check your local [`.kube/config`](https://blog.marcnuri.com/where-is-my-default-kubeconfig-file), the [Pod mounts](https://kubernetes.io/docs/tasks/run-application/access-api-from-pod/), and a few other places in order to infer your configuration. You can closely inspect the  [`Config.autoConfigure`](https://github.com/fabric8io/kubernetes-client/blob/23db4301478e74dac7154af70907bf98f2936ff5/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/Config.java#L273) to understand the flow.

So if you're running the client from a Pod, in a vanilla K8s cluster, you shouldn't need to provide any additional configuration settings since it should be automatically inferred.
